File: 3970-shortest-path-with-at-most-k-consecutive-identical-characters/3970-shortest-path-with-at-most-k-consecutive-identical-characters.cpp
#define tt tuple<int,int,int>
class Solution {
public:
    const int INF = 1e9;
    int shortestPath(int n, vector<vector<int>>& edges, string labels, int k) {
        vector<vector<pair<int,int>>>graph(n);
        for(auto const&itr: edges){
            graph[itr[0]].emplace_back(itr[1],itr[2]);
        }
        priority_queue<tt,vector<tt>,greater<tt>> pq;
        vector<vector<int>>dist(n,vector<int>(k+1,INF));
        pq.emplace(0,0,1);
        dist[0][1] = 0;
        while(!pq.empty()){
            auto [curr_dist, u,curr_consecutive] = pq.top();
            pq.pop();
            if(curr_dist>dist[u][curr_consecutive]) continue;
            if(u==n-1) return curr_dist;
            for(const auto& neghbour: graph[u]){
                int v = neghbour.first;
                int weight = neghbour.second;
                int next_consecutive = (labels[u]==labels[v])? curr_consecutive+1:1;
                if(next_consecutive>k) continue;
                if(curr_dist+weight<dist[v][next_consecutive]){
                    dist[v][next_consecutive] = curr_dist+weight;
                    pq.emplace(dist[v][next_consecutive],v,next_consecutive);
                }
            }
        }
        return -1;
    }
};

File: 3970-shortest-path-with-at-most-k-consecutive-identical-characters/3970-shortest-path-with-at-most-k-consecutive-identical-characters.py
import heapq
from typing import List

class Solution:
    def shortestPath(self, n: int, edges: List[List[int]], labels: str, k: int) -> int:
        adj = [[] for _ in range(n)]
        for u, v, w in edges:
            adj[u].append((v, w))
            
        dist = [[float('inf')] * (k + 1) for _ in range(n)]
        pq = []
        
        dist[0][1] = 0
        heapq.heappush(pq, (0, 0, 1))
        
        while pq:
            curr_dist, u, curr_consec = heapq.heappop(pq)
            
            if curr_dist > dist[u][curr_consec]:
                continue
                
            if u == n - 1:
                return curr_dist
                
            for v, weight in adj[u]:
                if labels[u] == labels[v]:
                    next_consec = curr_consec + 1
                else:
                    next_consec = 1
                
                if next_consec > k:
                    continue
                    
                if curr_dist + weight < dist[v][next_consec]:
                    dist[v][next_consec] = curr_dist + weight
                    heapq.heappush(pq, (dist[v][next_consec], v, next_consec))
                    
        return -1